### PR TITLE
Demonstrate a hybridized Raviart-Thomas method

### PR DESCRIPTION
**Pull request by lindsayad**
@nmnobre's RT addition could not have been more timely as I started an LDRD on hybridized methods this month. Hybridization of RT was the first example of a hybridized method dating back to 1965 but really canonicalized in Arnold's and Brezzi's "Mixed and nonconforming finite element methods: implementation, postprocessing, and error estimates".

Some details:

### Reduced global system size

For `TRI6`:
- non-hybridized with `grid_size=32`: 5185 dofs
- hybridized with `grid_size=32`: 3136 dofs

### Proper error convergence
```
lindad@pop-os:~/libmesh-devel/build/examples/vector_fe/vector_fe_ex7(raviart-thomas-hybridized)$ ./example-devel grid_size=2
L2 error is: 0.972572
HDiv semi-norm error is: 2.43859
HDiv error is: 2.62538
lindad@pop-os:~/libmesh-devel/build/examples/vector_fe/vector_fe_ex7(raviart-thomas-hybridized)$ ./example-devel grid_size=4
L2 error is: 0.500032
HDiv semi-norm error is: 1.27137
HDiv error is: 1.36617
lindad@pop-os:~/libmesh-devel/build/examples/vector_fe/vector_fe_ex7(raviart-thomas-hybridized)$ ./example-devel grid_size=8
L2 error is: 0.251364
HDiv semi-norm error is: 0.643384
HDiv error is: 0.690743
lindad@pop-os:~/libmesh-devel/build/examples/vector_fe/vector_fe_ex7(raviart-thomas-hybridized)$ ./example-devel grid_size=16
L2 error is: 0.125855
HDiv semi-norm error is: 0.322662
HDiv error is: 0.346338
```

### SPD matrix

```
lindad@pop-os:~/libmesh-devel/build/examples/vector_fe/vector_fe_ex7(raviart-thomas-hybridized)$ ./example-devel grid_size=3 -ksp_view_pmat
Mat Object: () 1 MPI process
  type: seqaij
row 0: (0, 8.)  (1, 0.)  (2, -2.)  (3, -2.)  (4, 0.) 
row 1: (0, 0.)  (1, 0.666667)  (2, 0.) 
row 2: (0, -2.)  (1, 0.)  (2, 4.)  (7, -2.)  (8, 2.22045e-16) 
row 3: (0, -2.)  (3, 4.)  (4, 0.)  (13, 2.22045e-16)  (14, -2.) 
row 4: (0, 0.)  (3, 0.)  (4, 0.666667) 
row 5: (5, 0.666667)  (6, 0.)  (7, 0.) 
row 6: (5, 0.)  (6, 4.)  (7, -2.)  (10, -2.)  (12, 5.30825e-16) 
row 7: (2, -2.)  (5, 0.)  (6, -2.)  (7, 8.)  (8, -2.) 
row 8: (2, 2.22045e-16)  (7, -2.)  (8, 4.)  (17, -2.)  (18, 0.) 
row 9: (9, 0.666667)  (10, 0.)  (11, 0.) 
row 10: (6, -2.)  (9, 0.)  (10, 8.)  (11, 0.)  (12, -2.) 
row 11: (9, 0.)  (10, 0.)  (11, 0.666667) 
row 12: (6, -6.82325e-17)  (10, -2.)  (12, 4.)  (20, 0.)  (21, -2.) 
row 13: (3, 8.88178e-16)  (13, 4.)  (14, -2.)  (17, -2.)  (19, 0.) 
row 14: (3, -2.)  (13, -2.)  (14, 8.)  (15, -2.)  (16, 0.) 
row 15: (14, -2.)  (15, 4.)  (16, 0.)  (23, 4.44089e-16)  (24, -2.) 
row 16: (14, 0.)  (15, 0.)  (16, 0.666667) 
row 17: (8, -2.)  (13, -2.)  (17, 8.)  (18, -2.)  (19, -2.) 
row 18: (8, 5.27356e-16)  (17, -2.)  (18, 4.)  (21, -2.)  (22, -6.66134e-16) 
row 19: (13, -2.22045e-16)  (17, -2.)  (19, 4.)  (27, 0.)  (28, -2.) 
row 20: (12, 0.)  (20, 0.666667)  (21, 0.) 
row 21: (12, -2.)  (18, -2.)  (20, 0.)  (21, 8.)  (22, -2.) 
row 22: (18, -8.88178e-16)  (21, -2.)  (22, 4.)  (30, 0.)  (31, -2.) 
row 23: (15, 2.22045e-16)  (23, 4.)  (24, -2.)  (28, -2.)  (29, 0.) 
row 24: (15, -2.)  (23, -2.)  (24, 8.)  (25, 0.)  (26, 0.) 
row 25: (24, 0.)  (25, 0.666667)  (26, 0.) 
row 26: (24, 0.)  (25, 0.)  (26, 0.666667) 
row 27: (19, -2.22045e-16)  (27, 4.)  (28, -2.)  (31, -2.)  (32, 0.) 
row 28: (19, -2.)  (23, -2.)  (27, -2.)  (28, 8.)  (29, 0.) 
row 29: (23, 0.)  (28, 0.)  (29, 0.666667) 
row 30: (22, 0.)  (30, 0.666667)  (31, 0.) 
row 31: (22, -2.)  (27, -2.)  (30, 0.)  (31, 8.)  (32, 0.) 
row 32: (27, 0.)  (31, 0.)  (32, 0.666667) 
L2 error is: 0.662752
HDiv semi-norm error is: 1.67467
HDiv error is: 1.80105

lindad@pop-os:~/libmesh-devel/build/examples/vector_fe/vector_fe_ex7(raviart-thomas-hybridized)$ ./example-devel grid_size=32 -ksp_type cg -pc_type hypre -pc_hypre_type boomeramg -ksp_monitor -options_left
  0 KSP Residual norm 2.645088819340e+01 
  1 KSP Residual norm 7.987205972185e-01 
  2 KSP Residual norm 7.124388216167e-03 
  3 KSP Residual norm 1.198453906227e-04 
  4 KSP Residual norm 1.811216108281e-06 
  5 KSP Residual norm 1.910616077348e-08 
  6 KSP Residual norm 3.094916518114e-10 
  7 KSP Residual norm 4.273815119150e-12 
L2 error is: 0.0629499
HDiv semi-norm error is: 0.16145
HDiv error is: 0.173289
```

### CODE REVIEW

**Comment by lindsayad** [comment]
I hope to add the postprocessing step which should boost the L2 error of the scalar solution by an order

**Comment by lindsayad** [comment]
woopsie, looks like I need to fix my autotools version

File: examples/vector_fe/vector_fe_ex7/vector_fe_ex7.C [hunk 1]
```diff
@@ -0,0 +1,551 @@
...
+    const auto lambda_n_dofs = lambda_dof_indices.size();
+
+    vector_fe->reinit(elem);
+    scalar_fe->reinit(elem);
+
+    libmesh_assert_equal_to(vector_n_dofs, vector_phi.size());
+    libmesh_assert_equal_to(scalar_n_dofs, scalar_phi.size());
+
+    // vector equation
+    A.resize(vector_n_dofs, vector_n_dofs);
```
Comment (jwpeterson) [inline examples/vector_fe/vector_fe_ex7/vector_fe_ex7.C:336]: I think there's
```
A = MatrixXd::Zero(vector_n_dofs, vector_n_dofs);
```
if you want to drop those lambdas.

Comment (jwpeterson) [reply examples/vector_fe/vector_fe_ex7/vector_fe_ex7.C:336]: Actually, I think even more appropriate would be:
```
A.setZero(vector_n_dofs, vector_n_dofs);
```

Comment (lindsayad) [reply examples/vector_fe/vector_fe_ex7/vector_fe_ex7.C:336]: Oh nice

File: examples/vector_fe/vector_fe_ex7/vector_fe_ex7.C [hunk 2]
```diff
@@ -0,0 +1,551 @@
...
+// The exact solution and error computation.
+#include "libmesh/exact_solution.h"
+#include "libmesh/enum_norm_type.h"
+#include "solution_function.h"
+
+// I/O utilities.
+#include "libmesh/getpot.h"
+#include "libmesh/exodusII_io.h"
+
+#include <Eigen/Dense>
```
Comment (roystgnr) [inline examples/vector_fe/vector_fe_ex7/vector_fe_ex7.C:58]: Let's get an `#ifdef LIBMESH_HAVE_EIGEN_DENSE` here somewhere?  I know everybody configures it on but we're at least still trying to keep it as an optional requirement.

File: examples/vector_fe/vector_fe_ex7/vector_fe_ex7.C [hunk 3]
```diff
@@ -0,0 +1,551 @@
...
+
+using namespace libMesh;
+using namespace Eigen;
+
+void fe_assembly(EquationSystems & es, bool global_solve);
+void assemble_divgrad(EquationSystems & es, const std::string & system_name);
+
```
Comment (roystgnr) [inline examples/vector_fe/vector_fe_ex7/vector_fe_ex7.C:65]: Could we get a thorough comment here with both the strong form and the weak form of the problem spelled out?

**Review by roystgnr — CHANGES_REQUESTED** [review]
I know you still have this marked as "Draft" but it's cool and I couldn't resist skimming through it.

This would benefit from a lot more comments in the new example in general, but I think the lack of a high-level comment and the configuration-proofing are my only big complaints.

**Comment by lindsayad** [comment]
Alright, enhanced convergence via postprocessing demonstrated!
```
lindad@pop-os:~/libmesh-devel/build/examples/vector_fe/vector_fe_ex7(raviart-thomas-hybridized)$ ./example-devel grid_size=2
L2 error is: 0.990879
HDiv semi-norm error is: 2.3968
HDiv error is: 2.59354
L2 error for p is: 0.493211
L2 error p_enriched is: 0.262107
lindad@pop-os:~/libmesh-devel/build/examples/vector_fe/vector_fe_ex7(raviart-thomas-hybridized)$ ./example-devel grid_size=4
L2 error is: 0.502064
HDiv semi-norm error is: 1.26741
HDiv error is: 1.36323
L2 error for p is: 0.257939
L2 error p_enriched is: 0.0709157
lindad@pop-os:~/libmesh-devel/build/examples/vector_fe/vector_fe_ex7(raviart-thomas-hybridized)$ ./example-devel grid_size=8
L2 error is: 0.251635
HDiv semi-norm error is: 0.642863
HDiv error is: 0.690358
L2 error for p is: 0.130415
L2 error p_enriched is: 0.0180657
lindad@pop-os:~/libmesh-devel/build/examples/vector_fe/vector_fe_ex7(raviart-thomas-hybridized)$ ./example-devel grid_size=16
L2 error is: 0.12589
HDiv semi-norm error is: 0.322593
HDiv error is: 0.346287
L2 error for p is: 0.0653893
L2 error p_enriched is: 0.00453757
lindad@pop-os:~/libmesh-devel/build/examples/vector_fe/vector_fe_ex7(raviart-thomas-hybridized)$ ./example-devel grid_size=32
L2 error is: 0.062954
HDiv semi-norm error is: 0.161442
HDiv error is: 0.173283
L2 error for p is: 0.0327174
L2 error p_enriched is: 0.00113572
```

**Comment by lindsayad** [comment]
> high-level comment

See [2df1d6a](https://github.com/libMesh/libmesh/pull/3668/commits/2df1d6a809b9543052b367968d7f90683657da81). I do not consider myself an expert on finite element notation, so please let me know if anything I've written is unclear or doesn't make sense

File: examples/vector_fe/vector_fe_ex7/vector_fe_ex7.C [hunk 4]
```diff
@@ -15,6 +15,87 @@
...
+// e.g. we no longer require that the normal component of the RT space be
+// continuous across interelement boundaries. In the notation of Arnold and
+// Brezzi this means a change from RT_0^k -> RT_{-1}^k where k is the polynomial
+// order of the basis (note that here k denotes the polynomial order of the
+// *divergence* of the vector shape functions, so for k = 0, the RT basis in
+// libMesh is FIRST). This breakage in continuity is accomplished by changing
+// from an FEFamily of RAVIART_THOMAS to L2_RAVIART_THOMAS. Instead of being
+// enforced by the finite element space itself, continuity of the normal
+// component of the vector field is enforced through Lagrange multipliers that
+// live on the skeleton of the mesh. Let's introduce a subspace of L2, M_{-1}^k,
```
Comment (roystgnr) [inline examples/vector_fe/vector_fe_ex7/vector_fe_ex7.C:43]: Is "skeleton" the term they use?  If so then let's keep it here, but I'd just have said "the sides of the elements", for `SIDE_HIERARCHIC`.  "Skeleton" makes me think of simplicial complexes and CW complexes, but technically the element sides here aren't *quite* a skeleton, they're the difference between the (d-1)-skeleton (the element sides, where `SIDE_HIERARCHIC` is continuous and well-defined) and the (d-2) skeleton (the vertices and in 3D the edges, where it isn't).

File: examples/vector_fe/vector_fe_ex7/vector_fe_ex7.C [hunk 5]
```diff
@@ -15,6 +15,87 @@
...
+//  and M_{-1}^k (E0_h). We will denote boundary faces by E. The hybridized
+// problem can be summarized as follows:
+//
+// find the approximate solutions (u_h, p_h, lambda_h) in
+// RT_{-1}^k x M_{-1}^k (I_h) x M^{-1}^k (E0_h)
+// such that
+//
+// (u, tau) - (p, div(tau)) - <lambda, tau*n>_E0 = -<g, tau*n>_E for all tau in RT_{-1}^k (I_h)
+// -(v, div(u)) = (f, v) for all v in M_{-1}^k (I_h)
+// -<mu, u*n> = 0 for all mu in M_{-1}^k (E0_h)
```
Comment (roystgnr) [inline examples/vector_fe/vector_fe_ex7/vector_fe_ex7.C:56]: n here is the normal vector?  If so are we making that clear?

File: examples/vector_fe/vector_fe_ex7/vector_fe_ex7.C (hunk 1 above)
Comment (jwpeterson) [reply examples/vector_fe/vector_fe_ex7/vector_fe_ex7.C:419]: Sorry to belabor the minor point, but I believe `setZero(m,n)` handles the resizing, so you don't need both `resize()` and `setZero()`. The following test code works as expected for me anyway:
```
  Eigen::MatrixXd A;
  A.setZero(3, 10);
  libMesh::out << "A.rows() = " << A.rows() << std::endl;
  libMesh::out << "A.cols() = " << A.cols() << std::endl;
  libMesh::out << A << std::endl;
```

File: examples/vector_fe/vector_fe_ex7/vector_fe_ex7.C (hunk 4 above)
Comment (lindsayad) [reply examples/vector_fe/vector_fe_ex7/vector_fe_ex7.C:43]: I have heard skeleton as the term used in the single hybridized discontinuous Galerkin presentation I've seen 😆

File: examples/vector_fe/vector_fe_ex7/vector_fe_ex7.C (hunk 1 above)
Comment (lindsayad) [reply examples/vector_fe/vector_fe_ex7/vector_fe_ex7.C:419]: I feel so assaulted 😄

File: examples/vector_fe/vector_fe_ex7/vector_fe_ex7.C (hunk 3 above)
Comment (lindsayad) [reply examples/vector_fe/vector_fe_ex7/vector_fe_ex7.C:156]: hopefully spelled out in the high level description now

**Review by roystgnr — APPROVED** [review]
We're going to need a MOOSE PR to fix that unrelated test regression, but that recipe went through previously with nothing but a weird also-unrelated Python failure.  I think this is ready to go.

**Review by jwpeterson — APPROVED** [review]

**Comment by lindsayad** [comment]
thanks for the reviews!

**Pull request merged**